blockchain/chain/src/store/chain_store.rs

```diff
@@ -515,6 +496,37 @@ where
     }
 }
 
+pub(crate) type TipsetCache = RwLock<LruCache<TipsetKeys, Arc<Tipset>>>;
+
+pub(crate) async fn tipset_from_keys<BS>(
+    cache: &TipsetCache,
+    store: &BS,
+    tsk: &TipsetKeys,
+) -> Result<Arc<Tipset>, Error>
+where
+    BS: BlockStore + Send + Sync + 'static,
+{
+    if let Some(ts) = cache.write().await.get(tsk) {
```

Comment:
why not use `cache.read()` here instead of getting a write lock?

Reply:
Because it's a least recently used (LRU) cache, and it needs to update the position of the entry on read to function correctly, which requires mutable access. There is a method called `peek` on this cache, which only requires immutable access, but that defeats the purpose of this cache.

Yes this limits the concurrent reads, but the lock is kept for a very short amount of time and this means that frequently accessed keys don't get evicted from the cache early